### ISSUE TITLE
[avro4s & Flink] How to create generic keyed serializer / deserializer schema

### ISSUE THREAD

**Issue by ngibanel** [sksamuel/avro4s#194]
Hi everyone,

I'm trying to create a generic keyed (Tuple2) serializer / deserializer schema and use it with the FlinkKafkaProducer / FlinkKafkaConsumer.

I have a KeyedSchema that implements the KeyedSerializationSchema and KeyedDeserializationSchema interface and use the GenericAvroSerde (#115)

```scala
import com.sksamuel.avro4s.{FromRecord, SchemaFor, ToRecord}
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.util.serialization.{KeyedDeserializationSchema, KeyedSerializationSchema}

import scala.reflect.ClassTag

class KeyedSchema[K : ClassTag : TypeInformation : SchemaFor : ToRecord : FromRecord, V : TypeInformation : ClassTag : SchemaFor : ToRecord : FromRecord](topic: String)(implicit val keyType: TypeInformation[K], implicit val valueType: TypeInformation[V])
  extends KeyedSerializationSchema[(K, V)] with KeyedDeserializationSchema[(K, V)] {

  lazy val keyAvroGenericSerde: GenericAvroSerde[K] = GenericAvroSerde[K]
  lazy val valueAvroGenericSerde: GenericAvroSerde[V] = GenericAvroSerde[V]

  override def serializeKey(element: (K, V)): Array[Byte] = keyAvroGenericSerde.serializer().serialize(topic, element._1)
  override def serializeValue(element: (K, V)): Array[Byte] = valueAvroGenericSerde.serializer().serialize(topic, element._2)
  override def getTargetTopic(element: (K, V)): String = topic
  override def isEndOfStream(nextElement: (K, V)): Boolean = false
  override def deserialize(messageKey: Array[Byte], message: Array[Byte], topic: String, partition: Int, offset: Long): (K, V) = (keyAvroGenericSerde.deserializer().deserialize(topic, messageKey), valueAvroGenericSerde.deserializer().deserialize(topic, message))
  override def getProducedType: TypeInformation[(K, V)] = createTuple2TypeInformation(keyType, valueType)
}
```

When I do a simple test,

```scala
import com.example.utils.KeyedSchema

import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.connectors.kafka.{FlinkKafkaProducer011 => FlinkKafkaProducer}

case class A(id: Int, label: String)
case class AKey(id: Int)

object Test {
  def main(args: Array[String]): Unit = {
    // Set up streaming execution environment
    val environment : StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    environment.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    environment.getConfig.disableSysoutLogging()

    // FlinkKafkaProducer

    val topic = "topic-test"

    implicit lazy val keyType: TypeInformation[AKey] = createTypeInformation[AKey]
    implicit lazy val valueType: TypeInformation[A] = createTypeInformation[A]

    @transient lazy val schema = new KeyedSchema[AKey, A](topic)
    lazy val producer: FlinkKafkaProducer[(AKey, A)] = new FlinkKafkaProducer("10.231.226.223:1026", topic, schema)

    environment
      .fromCollection(Seq((AKey(1), A(1, "A1"))))
      .addSink(producer)

    environment.execute("Generic Keyed Serializer / Deserializer Schema")
  }

}
```

I got the following exception when Flink try to serialize my KeyedSchema class :
```
Exception in thread "main" org.apache.flink.api.common.InvalidProgramException: Object com.example.utils.KeyedSchema@793be5ca is not serializable
	at org.apache.flink.api.java.ClosureCleaner.ensureSerializable(ClosureCleaner.java:108)
	at org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer011.<init>(FlinkKafkaProducer011.java:491)
	at org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer011.<init>(FlinkKafkaProducer011.java:444)
	at org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer011.<init>(FlinkKafkaProducer011.java:352)
Caused by: java.io.NotSerializableException: com.sksamuel.avro4s.FromValue$IntFromValue$
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.apache.flink.util.InstantiationUtil.serializeObject(InstantiationUtil.java:447)
	at org.apache.flink.api.java.ClosureCleaner.ensureSerializable(ClosureCleaner.java:106)
	... 7 more
```

Did you ever serialize a case class with avro4s and that works with a Flink Kafka Producer ?

Thank you for your help.

**Comment by wzorgdrager** [comment]
Hi,

I'm also struggling with this issue and haven't found a solution yet. The problem is that Flink needs to serialize the KeyedSchema class but since it has context bounds for `ToRecord`, `SchemaFor` and `FromRecord` (which aren't serializable), the KeyedSchema isn't serializable.

Hope someone has some useful suggestions.

**Comment by sksamuel** [comment]
Would making ToRecord etc serializeable fix the issue?

**Comment by wzorgdrager** [comment]
I'm not sure, currently it fails because this:
`com.sksamuel.avro4s.FromValue$StringFromValue` isn't serializable. I thought ToRecord and FromRecord already extended serializable? I think FromValue probably should extend it?

**Comment by wzorgdrager** [comment]
It will fix my issue partly. SchemaFor will probably never be serializable due to `org.apache.avro.Schema$RecordSchema` not being serializable which is out of your control. However, making FromRecord serializable would already be great since the conversion from GenericRecord to Scala (case) class is essential for me. Thanks!

**Comment by ngibanel** [comment]
Indeed the `org.apache.avro.Schema$RecordSchema` is not serializable and so SchemaFor will never be serializable. A workaround could be to serialize the class case in a map transformation and use a simple bytes serialization / deserialization schema.

**Comment by wzorgdrager** [comment]
Yes, well I got it both working now using just the Apache Avro library and a serializable FromRecord (I just edited avro4s). This is my deserialize method:

```scala
  override def deserialize(message: Array[Byte]): T = {
    var schema: Schema = null

    if (classOf[SpecificRecordBase].isAssignableFrom(inputClassType)) {
      schema = inputClassType.newInstance.asInstanceOf[SpecificRecordBase].getSchema
    } else {
      schema = ReflectData.get().getSchema(inputClassType);
    }

    val datumReader = new GenericDatumReader[GenericRecord](schema)
    val decoder = DecoderFactory.get().binaryDecoder(message, null)

    recordFrom(datumReader.read(null, decoder))
  }
```

recordFrom I passed implicitly (`implicit val recordFrom: FromRecord[T]`)

**Comment by sksamuel** [comment]
Do you want to PR the change?

**Comment by wzorgdrager** [comment]
Created the PR: #195

**Comment by sksamuel** [comment]
PR merged, will release 1.8.4 now.

**Comment by fubhy** [comment]
I ran into the same issue (KafkaSerializationSchema not serializable due to avro.schema). In case anyone runs into the same issue in the future... I worked around it a bit differently: Instead of deserializing/serializing right in the FlinkKafkaConsumer / FlinkKafkaProducer now, I offloaded it to a separate lambda:

```scala
package xxx

import java.util.Properties

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala.{StreamExecutionEnvironment, createTypeInformation}
import org.apache.flink.streaming.api.windowing.assigners.EventTimeSessionWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.connectors.kafka.{FlinkKafkaConsumer, FlinkKafkaProducer}
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.producer.ProducerConfig
import xxx.avro.{DummyKafkaSerde, GenericAvroSerde}
import xxx.records.{Probe, ProbeAggregate}
import xxx.streaming.ProbeAggregateFunction

object ProbeAggregator {
  private[this] val kafkaUrl = "kafka.service.consul:9092"
  private[this] val groupId = "flink.ProbeAggregator"
  private[this] val probeSerde = GenericAvroSerde[Probe](classOf[Probe])
  private[this] val probeAggSerde = GenericAvroSerde[ProbeAggregate](classOf[ProbeAggregate])

  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    val consumerConfig = new Properties()
    consumerConfig.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaUrl)
    consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId)

    val producerConfig = new Properties()
    producerConfig.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaUrl)

    val consumer = new FlinkKafkaConsumer[Array[Byte]]("Probe", new DummyKafkaSerde, consumerConfig)
    consumer.setStartFromGroupOffsets()

    val producer = new FlinkKafkaProducer[Array[Byte]]("ProbeAggregate", new DummyKafkaSerde, consumerConfig)
    producer.setWriteTimestampToKafka(true)

    val probes = env
      .addSource(consumer)
      .map(value => probeSerde.deserialize(value))
      .assignTimestampsAndWatermarks(
        new BoundedOutOfOrdernessTimestampExtractor[Probe](Time.seconds(5)) {
          override def extractTimestamp(probe: Probe): Long = probe.timestamp
        }
      )

    probes.print()

    val aggs = probes
      .keyBy(probe => probe.fingerprint)
      .window(EventTimeSessionWindows.withGap(Time.seconds(15)))
//      .allowedLateness(Time.minutes(5))
      .aggregate(new ProbeAggregateFunction())

    aggs.print()
    aggs.map(value => probeAggSerde.serialize(value)).addSink(producer)

    env.execute("Product aggregator")
  }
}
```

The DummyKafkaSerde class just forwards the raw byte array, serialization/deserialization happens later.
